diff --git a/gobject/meson.build b/gobject/meson.build
index 8ad90ac..f21da39 100644
--- a/gobject/meson.build
+++ b/gobject/meson.build
@@ -176,9 +176,11 @@ libgobject_dep = declare_dependency(link_with : libgobject,
 meson.override_dependency('gobject-2.0', libgobject_dep)
 
-gobject_query = executable('gobject-query', 'gobject-query.c',
-  install : true,
-  install_dir : get_option('devbindir'),
-  install_tag : 'bin-devel',
-  dependencies : [libglib_dep, libgobject_dep, app_profile_dep])
+if host_system != 'wasi'
+  gobject_query = executable('gobject-query', 'gobject-query.c',
+    install : true,
+    install_dir : get_option('devbindir'),
+    install_tag : 'bin-devel',
+    dependencies : [libglib_dep, libgobject_dep, app_profile_dep])
+endif
 
 install_data('gobject_gdb.py',
